BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Authentication/TokenCredentialEnvelope.cs

```diff
@@ -24,15 +24,31 @@ public TokenCredentialEnvelope(TokenCredential tokenCredential)
         /// <inheritdoc/>
         public override string GetToken(CancellationToken cancellationToken = default)
         {
-            var accessToken = this.tokenCredential.GetToken(requestContext: this.tokenRequestContext, cancellationToken: cancellationToken);
-            return accessToken.Token;
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+                var accessToken = this.tokenCredential.GetToken(requestContext: this.tokenRequestContext, cancellationToken: cancellationToken);
```

Comment:
We should cache the accessToken in a field because they are expensive to create. They are usually valid for 24 hours.
Assume that clients are going to use the token right away (within a few minutes).
Use the expiration time (minus 5 minutes or so) on the token to decide when to invalidate the cache.

Reply:
The expectation is that the implementing Credential class would handle local caching. 
I explicitly asked this question to the Identity architect and I was told NOT to cache locally.

Reply:
Are you aware of any published best-practices in this space?

Reply:
I've been over the source code for ManagedIdentityCredential and there's no caching of the access token that I can see. In Azure.Core, AccessToken caching happens in a different layer -- in the BearerTokenAuthenticationPolicy that is added to an HttpPipeline. We're not using HttpPipeline here, so this code can't use that.

I imagine that the AccessTokenCache in BearerTokenAuthenticationPolicy is the best practice. It's here: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs#L176

I find it pretty hard to read that source code because of the TaskCompletionSource and background refresh. From what I gather: First time, obviously you have to wait to call GetToken. Cache the returned value. Compute a RefreshOn time by subtracting 5 minutes from the ExpiresOn value. On subsequent calls, if DateTime.UtcNow is after ExpiresOn, obviously you need to block and get a new token. If it's close to expiry (after RefreshOn), kick off a background task to refresh it, but try and use the old one. Maybe your request will fail if you're too slow, but on the next retry, you'll block for the background refresh to complete.

Having said that, I would exercise caution adding caching directly on TokenCredentialEnvelope. Add it on the caller side. The main reason I say that is that the Linux Profiler agent is going to be accessing the TokenCredentialEnvelope too. It will call GetToken to get a token to call into the profiler endpoint. Because of the agent architecture, we need that token to be fairly long-lived (~15 minutes).